docs/design/data_safety.md

```diff
@@ -0,0 +1,153 @@
+Data Safety in ICU4X
+====================
+
+At its core, ICU4X is a set of algorithms that map input data to human-ready outputs. Since a core value proposition of ICU4X is the ability to dynamically load data, we can't always assume that our data are valid and pristine. This doc covers how we think about this problem and handle it in ICU4X.
+
+## Resolution
```

Comment:
suggestion: insert a summary of the problem to which the resolution pertains

I think there ought to be some text in between the current introductory paragraph, which is extremely general, and this resolution text, which launches into the points of justification for a specific technical solution.

IIUC, the debate has been centered around a tradeoff.  Specifically: 1) we desire safety/validity guarantees about our data;  2) we want optimal performance.  For the sake of explanation of my understanding, taken to one total extreme, you could write your code to _always_ validate data when using it at runtime, regardless of the source and context.  And that has a performance cost when compared to the other extreme, in which you naively don't write any validation code and get optimal performance, at a cost of security/guarantees between the code output and the user.

I think this document is trying to find a compromise in between those 2 logical extremes.  And it's doing so by decoupling (or at least minimizing) the validation code from the necessary core algorithm code.  This is perhaps analogous to programming languages that provide checked and unchecked versions of arithmetic operations (ex: a whether a `+` operation gives you an overflow error (checked), or if it silently gives you garbage for example by wrapping around the number space (unchecked)).

If this understanding applies in anyway to characterizing the previous discussion and/or where this proposal is going towards, then these ideas belong before this Resolution section and its detailed contents.

Reply:
How's this?